ir/instr.cpp

```diff
@@ -3603,23 +3613,55 @@ void ShuffleVector::rauw(const Value &what, Value &with) {
 }
 
 void ShuffleVector::print(ostream &os) const {
-  os << getName() << " = shufflevector " << *v1 << ", " << *v2;
-  for (auto m : mask)
-    os << ", " << m;
+  const char *str = nullptr;
+  switch(op) {
+  case LLVMIR_ShufVec:   str = "shufflevector"; break;
+  case SSSE3_PShufB128:  str = "ssse3.pshuf.b.128"; break;
+  case AVX2_PShufB:      str = "avx2.pshuf.b"; break;
+  case AVX512_PShufB512: str = "avx512.pshuf.b.512"; break;
+  }
+  os << getName() << " = " << str << " " << *v1 << ", " << *v2;
+  if (op == LLVMIR_ShufVec)
+    for (auto m : mask)
+      os << ", " << m;
 }
 
 StateValue ShuffleVector::toSMT(State &s) const {
-  auto vty = v1->getType().getAsAggregateType();
+  auto vty = static_cast<const VectorType*>(v1->getType().getAsAggregateType());
   auto sz = vty->numElementsConst();
   vector<StateValue> vals;
 
-  for (auto m : mask) {
-    if (m >= 2 * sz) {
-      vals.emplace_back(UndefValue(vty->getChild(0)).toSMT(s).value, true);
-    } else {
-      auto *vect = &s[m < sz ? *v1 : *v2];
-      vals.emplace_back(vty->extract(*vect, m % sz));
+  auto x86_shuffle = [&]() {
+    for (unsigned i = 0; i < vty->numElementsConst(); ++i) {
+      auto mask = vty->extract(s[*v2], expr::mkUInt(i, 64));
```

Comment:
aggregate type has an extract with an unsigned as arg

Reply:
Okay.
Interestingly, syntactically it accepts variable mask, but it crashes at SelDag. I'll update this

Reply:
Eh, I found that these intrinsics can sometimes take variable masks :/
We need VectorType's extract.
See test/Transforms/InstCombine/X86/x86-pshufb.ll 's demanded_elts_insertion().